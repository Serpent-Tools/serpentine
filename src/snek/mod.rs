//! Implementation of the snek language.
//!
//! Snek is pretty simplistic.
//! You define nodes, and their connections.
//!
//! ```snek
//! Image("...") .> Exec("cargo install nextest") .= base_image;
//!
//! base_image > Exec("cargo nextest run") 'tests;
//! base_image > Exec("cargo clippy") 'clippy;
//!
//! base_image > !'tests !'clippy Exec("cargo build") .> File("/target/...") .> Export("./bin/app") 'export;
//!
//! !'export RESULT;
//! ```

mod ast;
mod compiler;
mod parser;
pub mod span;
mod tokenizer;

use std::path::PathBuf;

pub use compiler::{CompileResult, compile_graph};
use miette::Diagnostic;
use span::Span;
use thiserror::Error;

/// An error occurred while building the graph.
#[derive(Debug, Error, Diagnostic)]
pub enum CompileError {
    /// Couldnt read a file needed to compile the code
    #[error("Could not read file {file}")]
    #[diagnostic(code(file_error))]
    FileReading {
        /// The file that couldnt be read
        file: PathBuf,
        /// The io error that caused it
        #[source]
        inner: std::io::Error,
    },

    /// The tokenizer encountered a value it didnt know what to do with
    #[error("unknown character {char:?} encountered in source code")]
    #[diagnostic(code(parsing::unknown_char))]
    UnknownCharacter {
        /// The location of the character
        #[label("This character was not understood by the lexer")]
        location: Span,
        /// The character
        char: char,
    },

    /// The parser encountered something different from what it expected.
    #[error("Expected `{expected}`")]
    #[diagnostic(code(parsing::unexpected_token))]
    UnexpectedToken {
        /// The token that was expected
        expected: String,
        /// The token that was encountered instead
        got: String,
        /// The location of the offending token
        #[label("Got `{got}`")]
        location: Span,
    },

    /// Mismatched type Error
    #[error("Expected `{expected}` got `{got}`")]
    #[diagnostic(code(compiler::type_mismatch))]
    TypeMismatch {
        /// The expected type
        expected: String,
        /// The type we got
        got: String,
        /// The location of the offending type in the source code
        #[label("Has type `{got}`")]
        location: Span,
        /// The node it was arugments for
        #[label("In call to this node")]
        node: Span,
    },

    /// Argument mismatch
    #[error("Expected {expected} arguments got {got}")]
    #[diagnostic(code(compiler::argument_count))]
    #[diagnostic(help(
        "Remember that chaining counts as a argument, `... > Foo(1, 2)` has 3 arguments passed for example."
    ))]
    ArgumentCountMismatch {
        /// The expected number of arguments
        expected: usize,
        /// The number of arguments we got
        got: usize,
        /// The location of the node in the source code
        #[label("This node expects {expected} arguments")]
        location: Span,
    },

    /// A name wasn't found in scope
    #[error("'{ident}' not found in scope")]
    #[diagnostic(code(compiler::item_not_found))]
    ItemNotFound {
        /// The name that wasn't found
        ident: String,
        /// The location of the name
        #[label("Item with this name not found")]
        location: Span,
    },

    /// A name was found in scope but wasnt the expected kind
    /// (e.g. a node was found when a label was expected)
    #[error("Expected a {expected}")]
    #[diagnostic(code(compiler::wrong_item_kind))]
    WrongItemKind {
        /// The expected kind
        expected: &'static str,
        /// The actual kind
        got: &'static str,
        /// The location of the name
        #[label("This is a '{got}'")]
        location: Span,
    },

    /// Statement found in unexpected context.
    #[error("{stmt} not allowed in {context}")]
    #[diagnostic(code(compiler::invalid_statement))]
    #[diagnostic(help("Maybe you meant to use a `{maybe}` statement?"))]
    InvalidStatement {
        /// The statement that was invalid
        stmt: &'static str,
        /// The context it was found in
        context: &'static str,
        /// A possible alternative statement
        maybe: &'static str,
        /// The location of the statement
        #[label("This statement is not allowed here")]
        location: Span,
    },

    /// A return was missing.
    #[error("No return found.")]
    #[diagnostic(code(compiler::no_return))]
    ReturnNotFound {
        /// The function that was missing a return
        #[label("In this function.")]
        location: Span,
    },

    /// Two returns were found.
    #[error("Multiple returns")]
    #[diagnostic(code(compiler::double_return))]
    DoubleReturn {
        /// Where was the second return found.
        #[label("Second return after exsisting return.")]
        location: Span,
    },

    /// A error occurred while inling function
    #[error("Inlining function failed")]
    InlineError {
        /// The error in the function
        #[diagnostic_source]
        error: Box<dyn Diagnostic + Send + Sync>,
        /// Where the function call was
        #[label("In this call")]
        call: Span,
    },

    /// A scope item was overwritten
    #[error("Item '{ident}' already defined in this scope")]
    #[diagnostic(code(compiler::item_already_defined))]
    #[diagnostic(help(
        "Snek sementically speaking does not have a specified exeuction order, so shadowing is not allowed."
    ))]
    ShadowedName {
        /// The name that was shadowed
        ident: String,
        /// The location of the name
        #[label("This name is already defined in this scope")]
        location: Span,
    },

    /// Unhandled internal error.
    #[error("INTERNAL ERROR - this is a bug, please report it.\n{0}")]
    #[diagnostic(code(internal_error))]
    InternalError(String),
}

impl CompileError {
    /// Create a `CompileError::InternalError`, but panic in debug mode instead
    pub fn internal(msg: impl Into<String>) -> Self {
        let msg = msg.into();
        debug_assert!(false, "{msg}");
        Self::InternalError(msg)
    }
}
