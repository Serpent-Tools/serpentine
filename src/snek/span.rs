//! Spans represent a range in the code

use std::ops::{Deref, DerefMut};
use std::path::PathBuf;

use crate::snek::CompileError;

/// A virtual file that can hold multiple files for error reporting
/// But look like one big file to miette.
#[derive(Debug)]
pub struct VirtualFile {
    /// The actual files, with their content and paths.
    files: Vec<(PathBuf, String)>,
}

/// A unique identifier for a file in the virtual file system
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct FileId(usize);

impl VirtualFile {
    /// Create a new empty virtual file
    pub fn new() -> Self {
        Self { files: Vec::new() }
    }

    /// Add a new file to this virtual file, returning its `FileId`.
    pub fn push(&mut self, path: PathBuf, code: &str) -> FileId {
        let mut current_end = 0;
        for (name, file) in &self.files {
            if *name == path {
                return FileId(current_end);
            }

            current_end = current_end.saturating_add(file.len());
        }

        self.files.push((path, code.to_owned()));

        FileId(current_end)
    }
}

impl miette::SourceCode for VirtualFile {
    fn read_span<'this>(
        &'this self,
        span: &miette::SourceSpan,
        context_lines_before: usize,
        context_lines_after: usize,
    ) -> Result<Box<dyn miette::SpanContents<'this> + 'this>, miette::MietteError> {
        let mut current_start = 0_usize;

        for (name, file) in &self.files {
            if span.offset() < current_start.saturating_add(file.len()) {
                let local_span = miette::SourceSpan::from((
                    span.offset().saturating_sub(current_start),
                    span.len(),
                ));

                let local_content =
                    file.read_span(&local_span, context_lines_before, context_lines_after)?;
                let local_content_span = local_content.span();
                let content_span = miette::SourceSpan::from((
                    local_content_span.offset().saturating_add(current_start),
                    local_content_span.len(),
                ));

                return Ok(Box::new(miette::MietteSpanContents::new_named(
                    name.display().to_string(),
                    local_content.data(),
                    content_span,
                    local_content.line(),
                    local_content.column(),
                    local_content.line_count(),
                )));
            }

            current_start = current_start.saturating_add(file.len());
        }

        Err(miette::MietteError::OutOfBounds)
    }
}

/// A span between `start` and `end`
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Span {
    /// The id of the file this span is in
    pub file_id: FileId,
    /// Start of the span (inclusive)
    pub start: usize,
    /// End of the span (exclusive)
    pub end: usize,
}

impl Span {
    /// Create a new span
    pub fn new(file_id: FileId, start: usize, end: usize) -> Self {
        debug_assert!(start <= end, "start ({start}) bigger than end ({end})");
        Self {
            file_id,
            start,
            end,
        }
    }

    /// Return a dummy span for calling apis in the compiler with compiler generated values.
    /// That also take a span for error reporting.
    pub fn dummy() -> Self {
        Self {
            file_id: FileId(0),
            start: 0,
            end: 0,
        }
    }

    /// Join two spans
    pub fn join(self, other: Self) -> Self {
        debug_assert!(
            self.file_id == other.file_id,
            "Joined spans from different files"
        );

        Self {
            file_id: self.file_id,
            start: self.start.min(other.start),
            end: self.end.max(other.end),
        }
    }

    /// Return a `Spanned` of the value and this span
    pub fn with<T>(self, value: T) -> Spanned<T> {
        Spanned(value, self)
    }

    /// return a clone of the string this span represents from the input string.
    pub fn index_str(self, value: &str) -> Result<Box<str>, CompileError> {
        value
            .get(self.start..self.end)
            .ok_or_else(|| {
                CompileError::internal(format!(
                    "Failed to index string {value:?} with {}..{}",
                    self.start, self.end
                ))
            })
            .map(Into::into)
    }
}

impl From<Span> for miette::SourceSpan {
    fn from(val: Span) -> Self {
        miette::SourceSpan::new(
            val.start.saturating_add(val.file_id.0).into(),
            val.end.saturating_sub(val.start),
        )
    }
}

/// Store `T` and a `Span`
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub struct Spanned<T>(pub T, pub Span);

impl<T> Deref for Spanned<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
impl<T> DerefMut for Spanned<T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

impl<T> Spanned<T> {
    /// Return the span of the data
    pub fn span(&self) -> Span {
        self.1
    }

    /// Return the owned data
    pub fn take(self) -> T {
        self.0
    }
}
