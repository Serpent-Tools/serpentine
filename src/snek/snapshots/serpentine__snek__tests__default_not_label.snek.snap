---
source: src/snek/mod.rs
expression: err
---
  x Compile Error

Error: compiler::wrong_item_kind

  x Expected a value
   ,-[<cli>:1:1]
 1 | DEFAULT
   : ^^^|^^^
   :    `-- This is a 'node'
   `----
